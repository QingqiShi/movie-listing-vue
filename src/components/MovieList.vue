<template>
  <transition-group tag="div" name="flip-list" class="movie-list">
    <movie
      v-for="(movie, i) in movies"
      :key="`movie-${movie.id}`"
      :movie="movie"
      :tabindex="i"
    />
  </transition-group>
</template>

<script>
import { mapGetters } from "vuex";
import Movie from "@/components/Movie.vue";

export default {
  components: { Movie },
  computed: {
    ...mapGetters({
      movies: "filteredMovies"
    })
  }
};
</script>

<style lang="scss">
.movie-list {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));

  > .movie {
    transition: all 0.5s;
  }
}

.flip-list-enter,
.flip-list-leave-to {
  opacity: 0;
  transform: scale(0.7);
}

.flip-list-leave-active {
  position: absolute;
}
</style>
