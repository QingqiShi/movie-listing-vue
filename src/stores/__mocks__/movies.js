const state = {
  configuration: {
    imageBaseUrl: "https://images.com/",
    posterSizes: ["w100", "w300", "original"],
    backdropSizes: ["w100", "w300", "original"]
  },
  movies: [
    {
      id: 1000,
      title: "Avatar",
      genre_ids: [0],
      popularity: 230,
      vote_average: 5
    }
  ],
  genres: [{ id: 0, name: "Action" }, { id: 1, name: "Horror" }],
  sortOrder: false,
  sortProperty: "popularity",
  filterRating: 3,
  filterGenres: { 0: false, 1: true },
  error: ""
};

const getters = {
  /** @param {typeof state} state */
  ratingFilteredMovies: state => state.movies,
  availableGenres: () => [{ id: 0, name: "Action" }],
  filteredMovies: () => []
};

const mutations = {
  setConfiguration: jest.fn(),
  setMovies: jest.fn(),
  setGenres: jest.fn(),
  setSort: jest.fn(),
  setFilter: jest.fn(),
  setError: jest.fn()
};

const actions = {
  initialFetch: jest.fn(() => Promise.resolve()),
  fetchConfiguration: jest.fn(() => Promise.resolve()),
  fetchMovies: jest.fn(() => Promise.resolve()),
  fetchGenres: jest.fn(() => Promise.resolve())
};

export default { state, getters, mutations, actions };
