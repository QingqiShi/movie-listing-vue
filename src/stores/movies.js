import api from "@/constants/api";
import app from "@/constants/app";

const { CONFIGURATION, NOW_PLAYING, GENRES_LIST } = api;
const { LOCALE } = app;

const state = {
  configuration: {},
  /** @type {Movie[]} List of all movies */
  movies: [],
  /** @type {Genre[]} List of all genres */
  genres: [],
  /** @type {boolean} True for accending, false for decending */
  sortOrder: false,
  /** @type {"popularity"|"vote_average"} */
  sortProperty: "popularity",
  /** @type {number} Number between 0 and 10 */
  filterRating: 3,
  /** @type {{[id:number]: boolean}} */
  filterGenres: {},
  /** @type {string} */
  error: ""
};

const getters = {
  /** @type {(state:State) => Movie[]} */
  ratingFilteredMovies: state =>
    state.movies.filter(movie => movie.vote_average >= state.filterRating),

  /** @type {(state:State, getters:Getters) => Genre[]} */
  availableGenres: (state, getters) =>
    state.genres.filter(genre => {
      for (let movie of getters.ratingFilteredMovies) {
        if (movie.genre_ids.includes(genre.id)) {
          return true;
        }
      }
      return state.filterGenres[genre.id];
    }),

  /** @type {(state:State, getters:Getters) => Movie[]} */
  filteredMovies: (state, getters) => {
    const { availableGenres, ratingFilteredMovies } = getters;
    const { filterGenres } = state;

    const genres = availableGenres.filter(genre => filterGenres[genre.id]);
    if (!genres.length) return ratingFilteredMovies;

    return ratingFilteredMovies.filter(({ genre_ids }) =>
      genres.reduce((show, genre) => show && genre_ids.includes(genre.id), true)
    );
  }
};

const mutations = {
  /** @type {(state:State, payload:Configuration) => void} */
  setConfiguration(state, { imageBaseUrl, posterSizes, backdropSizes }) {
    state.configuration = { imageBaseUrl, posterSizes, backdropSizes };
  },

  /** @type {(state:State, movies:Movie[]) => void} */
  setMovies(state, movies) {
    state.movies = movies;
  },

  /** @type {(state:State, genres:Genre[]) => void} */
  setGenres(state, genres) {
    state.genres = genres;
    state.filterGenres = {};
    genres.forEach(genre => {
      state.filterGenres[genre.id] = false;
    });
  },

  /** @type {(state:State, payload:{property:SortProperty, order:boolean}) => void} */
  setSort(state, { property, order }) {
    state.sortOrder = order;
    state.sortProperty = property;
    state.movies = state.movies.sort((a, b) =>
      order ? a[property] - b[property] : b[property] - a[property]
    );
  },

  /** @type {(state:State, payload:{type:"rating"|"genres", value:number}) => void} */
  setFilter(state, { type, value }) {
    if (type === "rating") {
      state.filterRating = value;
    } else {
      const newFilters = { ...state.filterGenres };
      newFilters[value] = !state.filterGenres[value];
      state.filterGenres = newFilters;
    }
  },

  /** @type {(state:State, errorMsg:string) => void} */
  setError(state, errorMsg) {
    state.error = errorMsg;
  }
};

const actions = {
  /** @param {import('vuex').ActionContext} context */
  async initialFetch({ commit, dispatch }) {
    try {
      await Promise.all([
        dispatch("fetchConfiguration"),
        dispatch("fetchMovies"),
        dispatch("fetchGenres")
      ]);
      commit("setSort", { property: "popularity", order: false });
    } catch (e) {
      commit("setError", e);
    }
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchConfiguration({ commit }) {
    const result = await fetchApi(
      CONFIGURATION(),
      "FetchConfiguration",
      result => result && result.images
    );

    commit("setConfiguration", {
      imageBaseUrl: result.images.base_url,
      posterSizes: result.images.poster_sizes,
      backdropSizes: result.images.backdrop_sizes
    });
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchMovies({ commit }) {
    const result = await fetchApi(
      NOW_PLAYING(LOCALE),
      "FetchMovies",
      result => result && result.results
    );

    commit("setMovies", result.results);
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchGenres({ commit }) {
    const result = await fetchApi(
      GENRES_LIST(LOCALE),
      "FetchGenres",
      result => result && result.genres
    );

    commit("setGenres", result.genres);
  }
};

/**
 * @param {string} url
 * @param {string} name
 */
async function fetchApi(url, name, validator) {
  let result;

  try {
    const response = await fetch(url);
    result = await response.json();
  } catch {
    throw `${name} failed, check your network and try again`;
  }

  if (!validator(result)) {
    throw `${name} returned invalid result`;
  }

  return result;
}

export default { state, getters, mutations, actions };

/* Type definitions to help vscode do auto suggest */

/** @typedef {{id:number, name:string}} Genre */
/**
 * @typedef {object} Movie
 * @property {number} Movie.id
 * @property {string} Movie.title
 * @property {number} Movie.popularity
 * @property {number} Movie.vote_average
 * @property {number[]} Movie.genre_ids
 * @property {string} Movie.poster_path
 */
/**
 * @typedef {object} Configuration
 * @property {string} Configuration.imageBaseUrl
 * @property {string} Configuration.posterSizes
 * @property {string} Configuration.backdropSizes
 */
/** @typedef {typeof state} State */
/**
 * @typedef {object} Getters
 * @property {Movie[]} Getters.ratingFilteredMovies
 * @property {Genre[]} Getters.availableGenres
 * @property {Movie[]} Getters.filteredMovies
 */
/** @typedef {typeof state.sortProperty} SortProperty */
