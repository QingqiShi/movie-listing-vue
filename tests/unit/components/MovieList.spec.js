import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import MovieList from "@/components/MovieList.vue";
import movies from "@/stores/movies";

jest.mock("@/stores/movies");

const localVue = createLocalVue();
localVue.use(Vuex);

describe("MovieList.vue", () => {
  it("renders", () => {
    const store = new Vuex.Store({ modules: { movies } });
    const wrapper = shallowMount(MovieList, { store, localVue });
    expect(wrapper).toMatchSnapshot();
  });
});
