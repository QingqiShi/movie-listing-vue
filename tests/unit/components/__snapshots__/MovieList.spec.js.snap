// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList.vue renders 1`] = `<span tag="div" name="flip-list" class="movie-list"></span>`;
